### ISSUE TITLE
PERF102 does not trigger when the loop var is used previously

### ISSUE THREAD

**Issue by ThiefMaster** [astral-sh/ruff#8786]
```py
dict_a = {}
dict_b = {}

for loopvar in {}:
    print(dict_a[loopvar])

for loopvar, val in dict_b.items():
    print(val)

for newloopvar, val in dict_b.items():
    print(val)
```

```
$ ruff --isolated --select B007,PERF102 --no-cache rufftest/ruff_sample.py
rufftest/ruff_sample.py:7:5: B007 Loop control variable `loopvar` not used within loop body
rufftest/ruff_sample.py:10:5: B007 Loop control variable `newloopvar` not used within loop body
rufftest/ruff_sample.py:10:24: PERF102 When using only the values of a dict use the `values()` method
Found 3 errors.
No fixes available (2 hidden fixes can be enabled with the `--unsafe-fixes` option).
```

**Comment by zanieb** [comment]
Hm this seems to be a detail of the `SemanticModel::is_unused` implementation which will say a variable is used if it is shadowed. For example, this reproduces with the simpler:

```python
key = 1
print(key)

for key, val in {}.items():
    print(val)
```

**Comment by zanieb** [comment]
And.. I think we can maybe be considered correct here although it is confusing because `key` is mutated:

```python
key = 1
print(key)

for key, val in {"x": 1, "y": 2}.items():
    print(val)

print(key)
```
```
1
1
2
y
```

I guess we could check if it's used _after_ the loop but that feels like a separate rule?

**Comment by ThiefMaster** [comment]
Which, to be fair, is behavior nobody should ever rely upon ;)

TBH, maybe there should be a new rule that warns if you rely on the loop variable leaking outside the loop...

**Comment by mikeleppane** [comment]
I am working on this.

**Comment by dhruvmanila** [comment]
> With given example above, PERF102 didn't get line 7. I think that's the expected behavior.

Can you explain why? `PERF102` isn't triggered for the second loop even though the `loopvar` variable remains unused within the body of that for loop.
